README.md
# .prg Language README

Welcome to the vscode extension for the `.prg` language! This extension aims to make your experience with the `.prg` language smooth, efficient, and delightful. Learn more about the features and capabilities of this extension here: https://github.com/agbowman/.prg-formatter. More quality of life improvements are on the way.

## HOW TO USE

Download folder from https://github.com/agbowman/.prg-formatter to your desired location on your computer. Open a project containing .prg file. Go to extensions in vscode -> click on the three dots -> install from VSIX -> select the .vsix file from the folder you downloaded. Reload vscode. Open .prg file. If needed: Right click on the file -> Format Document.

## TO ADD MORE YOURSELF

to create your own extension with yo code:

npm install -g yo generator-code
yo code
Select New Language Support
Enter prg

or to update the the current language in this extension:

npm install vsce
add to customprg.tmLanguage.json or language-configuration.json
vsce package
new .vsix file will be created in the folder
install the .vsix file in vscode

## Features

The `.prg` language extension offers several visual aids to enhance your experience with the `.prg` language.

## Requirements

Ensure that you have the latest version of Visual Studio Code installed. No other special requirements or dependencies are needed for this extension.

## Extension Settings

This extension provides several settings to tailor your `.prg` experience:

- `.prgLanguage.autoFormat`: Enables automatic formatting of `.prg` files.
- `.prgLanguage.lintOnSave`: Checks your `.prg` file for errors every time you save.

## Known Issues

As of now, there are no known major issues. If you encounter any problems or have feature requests, kindly raise an issue on the GitHub repository.

## Release Notes

Stay updated with the changes and improvements:

### 1.0.0

Initial release of `.prg` language support.


CHANGELOG.md
# Change Log

All notable changes to the "prg" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

src/extension.ts
import * as vscode from 'vscode';
import { completionFunctions } from './completionItems';




const ifDecorationType = vscode.window.createTextEditorDecorationType({
  // Define the styles for the decoration
  borderRadius: '3px',
  backgroundColor: 'rgba(255, 255, 0, 0.3)' // Example color, change as needed
});

function createInitialsRegExp(completionFunctions: { label: string; }[]): RegExp {
  const uniqueInitials = new Set(completionFunctions.map(func => func.label[0].toLowerCase()));
  return new RegExp(`[${Array.from(uniqueInitials).join('')}]$`, 'i');
}

const initialsRegExp = createInitialsRegExp(completionFunctions);

class MyCompletionItemProvider implements vscode.CompletionItemProvider {
  public provideCompletionItems(
    document: vscode.TextDocument,
    position: vscode.Position,
    token: vscode.CancellationToken,
    context: vscode.CompletionContext
  ): vscode.CompletionItem[] | undefined {
    const linePrefix = document.lineAt(position).text.substr(0, position.character);
    if (!initialsRegExp.test(linePrefix)) {
      return undefined;
    }

    // Convert your imported completion functions to VS Code CompletionItems
    return completionFunctions.map(func => {
      const completionItem = new vscode.CompletionItem(func.label, func.kind);
      completionItem.documentation = new vscode.MarkdownString(func.documentation);
      completionItem.insertText = func.insertText;
      return completionItem;
    });
  }
}
function updateIfEndifDecorations(editor: vscode.TextEditor) {
  // Clear existing decorations
  editor.setDecorations(ifDecorationType, []);

  const text = editor.document.getText();
  const ifRegex = /if\b/gi;
  const endifRegex = /endif\b/gi;

  const ifRanges: vscode.Range[] = [];
  const endifRanges: vscode.Range[] = [];

  let match;
  while (match = ifRegex.exec(text)) {
      const startPos = editor.document.positionAt(match.index);
      const endPos = editor.document.positionAt(match.index + match[0].length);
      const range = new vscode.Range(startPos, endPos);
      ifRanges.push(range);
  }

  while (match = endifRegex.exec(text)) {
      const startPos = editor.document.positionAt(match.index);
      const endPos = editor.document.positionAt(match.index + match[0].length);
      const range = new vscode.Range(startPos, endPos);
      endifRanges.push(range);
  }

  // Assuming a simple structure where each 'if' matches the following 'endif'
  // This logic should be adjusted for nested or more complex structures
  for (let i = 0; i < Math.min(ifRanges.length, endifRanges.length); i++) {
      if (editor.selection.intersection(ifRanges[i]) || editor.selection.intersection(endifRanges[i])) {
          editor.setDecorations(ifDecorationType, [ifRanges[i], endifRanges[i]]);
          break;
      }
  }
}
// You would also need to register the provider with VSCode in the extension activation function.
// For example:
export function activate(context: vscode.ExtensionContext) {
  console.log('Extension "vscode prg-formatter" is now active.');

  const provider = new MyCompletionItemProvider();

  // We use the spread operator to convert the string into an array of characters
  const triggerCharacters = [...'abcdefghijklmnopqrstuvwxyz'];

  const providerDisposable = vscode.languages.registerCompletionItemProvider(
    'customprg', // The language identifier
    provider,
    ...triggerCharacters // Spread the array to use each letter as a trigger character
  );

  context.subscriptions.push(providerDisposable);



  //---now for the highlighter code:
      // Register the event listener for text editor selection changes
      context.subscriptions.push(vscode.window.onDidChangeTextEditorSelection(event => {
        updateIfEndifDecorations(event.textEditor);
    }));

    // Initial call to set decorations for the active editor
    if (vscode.window.activeTextEditor) {
        updateIfEndifDecorations(vscode.window.activeTextEditor);
    }
}

export function deactivate() {
  // Clean up if necessary
}


out/completionItems.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.completionFunctions = void 0;
const vscode = __importStar(require("vscode"));
exports.completionFunctions = [
    {
        label: 'AVG',
        kind: vscode.CompletionItemKind.Function,
        documentation: 'Returns the average of two or more values.',
        insertText: 'AVG'
    },
    {
        label: 'SUM',
        kind: vscode.CompletionItemKind.Function,
        documentation: 'Calculates the sum of a series of numbers.',
        insertText: 'SUM'
    }
    // ... other functions
];
//# sourceMappingURL=completionItems.js.map

out/extension.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.deactivate = exports.activate = void 0;
const vscode = __importStar(require("vscode"));
const completionItems_1 = require("./completionItems");
const ifDecorationType = vscode.window.createTextEditorDecorationType({
    // Define the styles for the decoration
    borderRadius: '3px',
    backgroundColor: 'rgba(255, 255, 0, 0.3)' // Example color, change as needed
});
function createInitialsRegExp(completionFunctions) {
    const uniqueInitials = new Set(completionFunctions.map(func => func.label[0].toLowerCase()));
    return new RegExp(`[${Array.from(uniqueInitials).join('')}]$`, 'i');
}
const initialsRegExp = createInitialsRegExp(completionItems_1.completionFunctions);
class MyCompletionItemProvider {
    provideCompletionItems(document, position, token, context) {
        const linePrefix = document.lineAt(position).text.substr(0, position.character);
        if (!initialsRegExp.test(linePrefix)) {
            return undefined;
        }
        // Convert your imported completion functions to VS Code CompletionItems
        return completionItems_1.completionFunctions.map(func => {
            const completionItem = new vscode.CompletionItem(func.label, func.kind);
            completionItem.documentation = new vscode.MarkdownString(func.documentation);
            completionItem.insertText = func.insertText;
            return completionItem;
        });
    }
}
function updateIfEndifDecorations(editor) {
    // Clear existing decorations
    editor.setDecorations(ifDecorationType, []);
    const text = editor.document.getText();
    const ifRegex = /if\b/gi;
    const endifRegex = /endif\b/gi;
    const ifRanges = [];
    const endifRanges = [];
    let match;
    while (match = ifRegex.exec(text)) {
        const startPos = editor.document.positionAt(match.index);
        const endPos = editor.document.positionAt(match.index + match[0].length);
        const range = new vscode.Range(startPos, endPos);
        ifRanges.push(range);
    }
    while (match = endifRegex.exec(text)) {
        const startPos = editor.document.positionAt(match.index);
        const endPos = editor.document.positionAt(match.index + match[0].length);
        const range = new vscode.Range(startPos, endPos);
        endifRanges.push(range);
    }
    // Assuming a simple structure where each 'if' matches the following 'endif'
    // This logic should be adjusted for nested or more complex structures
    for (let i = 0; i < Math.min(ifRanges.length, endifRanges.length); i++) {
        if (editor.selection.intersection(ifRanges[i]) || editor.selection.intersection(endifRanges[i])) {
            editor.setDecorations(ifDecorationType, [ifRanges[i], endifRanges[i]]);
            break;
        }
    }
}
// You would also need to register the provider with VSCode in the extension activation function.
// For example:
function activate(context) {
    console.log('Extension "vscode prg-formatter" is now active.');
    const provider = new MyCompletionItemProvider();
    // We use the spread operator to convert the string into an array of characters
    const triggerCharacters = [...'abcdefghijklmnopqrstuvwxyz'];
    const providerDisposable = vscode.languages.registerCompletionItemProvider('customprg', // The language identifier
    provider, ...triggerCharacters // Spread the array to use each letter as a trigger character
    );
    context.subscriptions.push(providerDisposable);
    //---now for the highlighter code:
    // Register the event listener for text editor selection changes
    context.subscriptions.push(vscode.window.onDidChangeTextEditorSelection(event => {
        updateIfEndifDecorations(event.textEditor);
    }));
    // Initial call to set decorations for the active editor
    if (vscode.window.activeTextEditor) {
        updateIfEndifDecorations(vscode.window.activeTextEditor);
    }
}
exports.activate = activate;
function deactivate() {
    // Clean up if necessary
}
exports.deactivate = deactivate;
//# sourceMappingURL=extension.js.map